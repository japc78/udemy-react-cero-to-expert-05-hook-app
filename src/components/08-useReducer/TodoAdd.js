import React from 'react'
import { useForm } from '../../hooks/useForm';

export const TodoAdd = ({ handleAddTodo }) => {
	// Al manejar solo un campo (input) se puede hacer destructuracion del objeto y extraer la descripcion.
	const [{ description }, handelInputChange, reset ] = useForm({
		description: '',
	});
	// console.log(description);


	// Manejador del formulario.
	const handleSubmit = (e) => {
		e.preventDefault();

		if (description.trim().length <= 1) {
			return;
		}

		const newTodo = {
			id: new Date().getTime(),
			desc: description,
			done: false
		};

		handleAddTodo(newTodo);
		reset();
	}

	return (
		<>
			<h4>Agregar TODO</h4>
			<hr/>

			<form onSubmit= { handleSubmit }>
				<input
					type="text"
					name="description"
					onChange={ handelInputChange }
					className="form-control"
					placeholder="Aprender ..."
					value={ description }
					autoComplete="off"/>

				<button
					type="submit"
					className="btn btn-sm btn-outline-primary mt-1 btn-block"
				>
					Agregar
				</button>
			</form>
		</>
	)
}
