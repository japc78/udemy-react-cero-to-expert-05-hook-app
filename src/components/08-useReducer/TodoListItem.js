import React from 'react'

export const TodoListItem = ( {todo, idx, handleDelete, handleToggle }) => {
	return (
		// todoListItem, argumentos: todo, index, handleDelete, handleToogle
		<li
			key={todo.id}
			className="list-group-item"
		>
			<p
				onClick={ () => handleToggle(todo.id) }
				className={ `${ todo.done && 'complete'}` }>
					{idx + 1}. {todo.desc}
			</p>
			<button
				onClick= { () => handleDelete(todo.id) }
				className="btn btn-sm btn-danger">
					borrar
			</button>

		</li>
	)
}
