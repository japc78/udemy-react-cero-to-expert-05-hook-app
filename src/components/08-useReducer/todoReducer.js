

export const todoReducer = (state = [], action) => {
	switch (action.type) {
		case 'add':
			return [ ...state, action.payload]

		case 'delete':
			//filter retorna un nuevo array.
			// En este caso devuleve un array con todos aquelos elementos que sean distintos al id pasado por el payload.
			return state.filter(todo => todo.id !== action.payload);

		// Forma larga
		case 'toggle-old':
			return state.map(todo => {
				if (todo.id === action.payload) {
					return {
						...todo,
						done: !todo.done
					}

				}else {
					return todo;
				}
			})


		// Forma corta con return implicito en el map y ternario.
		case 'toggle':
			return state.map( todo =>
				( todo.id === action.payload )
					? {...todo, done: !todo.done}
					: todo
			)

		default:
			return state;
	}
}