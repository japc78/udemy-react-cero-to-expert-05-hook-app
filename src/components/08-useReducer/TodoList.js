import React from 'react'
import { TodoListItem } from './TodoListItem'

export const TodoList = ( { todos, handleDelete, handleToggle } ) => {
	return (
		/* TodoList, argumetnos: todos, handleDelete, handleToogle*/
		<ul className="list-group list-group-flush">
			{
				todos.map((todo, idx) => (
					<TodoListItem
						key = {todo.id}
						todo = { todo }
						idx = { idx }
						handleDelete = { handleDelete }
						handleToggle = { handleToggle }
					/>
				))
			}
		</ul>
	)
}
