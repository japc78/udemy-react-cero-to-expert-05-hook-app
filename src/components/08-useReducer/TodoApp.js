import React, { useEffect, useReducer } from 'react'
import { todoReducer } from './todoReducer'
import { useForm } from '../../hooks/useForm'

import { TodoList } from './TodoList'
import { TodoAdd } from './TodoAdd'

import './styles.css'
export const TodoApp = () => {

	// Estado inicial, se le pasa al useReducer
	// const initialState = [{
	// 	id: new Date().getTime(),
	// 	desc: 'Aprender React',
	// 	done: false
	// }]

	const init = () => {
		// return [{
		// 	id: new Date().getTime(),
		// 	desc: 'Aprender React',
		// 	done: false
		// }]

		// Se recoge la informacion guardada en el localstore
		// Si JSON.parse(localStorage.getItem('todos')) es null retornara un array vacio.
		return JSON.parse(localStorage.getItem('todos')) || []
	}

	const [ todos, dispatch ] = useReducer(todoReducer, [], init);
	// console.log(todos);




	// Se utiliza useEffect para que cada vez que se actualize el estado todos, se grabe en localstore.
	// Se pasa a JSON con stringify los todos ya que localstore solo permite guardar en String.
	useEffect( () => {
		localStorage.setItem('todos', JSON.stringify(todos))
	}, [todos])


	const handleDelete = ( todoId ) => {
		console.log("Aqui: " + todoId);

		//Tarea
		// Crear la action
		const action = {
			type: 'delete',
			payload: todoId
		}

		// hacer el dispach
		dispatch(action);
	}

	const handleToggle = (todoId) => {
		dispatch({
			type: 'toggle',
			payload: todoId
		})
	}


	const handleAddTodo = ( newTodo ) => {
		dispatch({
			type: 'add',
			payload: newTodo
		})
	}

	return (
		<div>
			<h1>TodoApp ({ todos.length })</h1>
			<hr/>

			<div className="row">
				<div className="col-7">
					<TodoList
						todos = { todos }
						handleDelete = { handleDelete }
						handleToggle = { handleToggle }
					/>
				</div>

				<div className="col-5">
					<TodoAdd handleAddTodo = { handleAddTodo }/>
				</div>
			</div>
		</div>
	)
}
