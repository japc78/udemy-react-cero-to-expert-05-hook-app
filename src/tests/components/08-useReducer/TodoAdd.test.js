import React from 'react'
import '@testing-library/jest-dom'
import { shallow } from 'enzyme'
import { TodoAdd } from '../../../components/08-useReducer/TodoAdd'

describe('Test on <TodoAdd />', () => {

	const handleAddTodo = jest.fn();
	const wrapper = shallow(
		<TodoAdd
			handleAddTodo = { handleAddTodo }
		/>
	)

	test('should be shown correctly', () => {
		expect( wrapper ).toMatchSnapshot();
	});


	test('should not call the handleAddTodo', () => {
		const formSubmit = wrapper.find('form').prop('onSubmit');
		//Se comprueba que formSubmit es igual a una funcion
		// console.log(formSubmit);

		formSubmit({ preventDefault(){}});
		expect( handleAddTodo ).toHaveBeenCalledTimes(0);
	});


	test('should call the handleAddTodo function with one args', () => {
		const value = 'New Tast';

		wrapper.find('input').simulate('Change', {
			 target: {
				value,
				name: 'description'
			}
		});
		const formSubmit = wrapper.find('form').prop('onSubmit');
		formSubmit({ preventDefault(){}});

		expect( handleAddTodo ).toHaveBeenCalledTimes(1);
		expect( handleAddTodo ).toHaveBeenCalledWith(expect.any(Object));
		expect( handleAddTodo ).toHaveBeenCalledWith({
			id: expect.any(Number),
			desc: value,
			done: false
		});

		// Se comprueba que se realiza el reset
		expect( wrapper.find('input').prop('value')).toBe("");

	})
})
