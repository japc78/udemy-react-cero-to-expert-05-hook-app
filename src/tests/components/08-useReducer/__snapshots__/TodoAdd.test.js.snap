// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on <TodoAdd /> should be shown correctly 1`] = `
<Fragment>
  <h4>
    Agregar TODO
  </h4>
  <hr />
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="form-control"
      name="description"
      onChange={[Function]}
      placeholder="Aprender ..."
      type="text"
      value=""
    />
    <button
      className="btn btn-sm btn-outline-primary mt-1 btn-block"
      type="submit"
    >
      Agregar
    </button>
  </form>
</Fragment>
`;
